tests/test_cluster.py

```diff
@@ -0,0 +1,628 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the :mod:`pennylane.cluster` submodule.
+"""
+from collections.abc import Sequence
+
+import pytest
+from pennylane import numpy as np
+
+import pennylane as qml
+
+
+try:
+    import torch
+except ImportError as e:
+    pass
+
+
+try:
+    import tensorflow as tf
+
+    if tf.__version__[0] == "1":
+        print(tf.__version__)
+        import tensorflow.contrib.eager as tfe
+
+        tf.enable_eager_execution()
+        Variable = tfe.Variable
+    else:
+        from tensorflow import Variable
+except ImportError as e:
+    pass
+
+
+@pytest.fixture
+def qnodes(interface):
+    """fixture returning some QNodes"""
+    dev1 = qml.device("default.qubit", wires=2)
+    dev2 = qml.device("default.qubit", wires=2)
+
+    @qml.qnode(dev1, interface=interface)
+    def qnode1(x):
+        qml.RX(x[0], wires=0)
+        qml.RY(x[1], wires=0)
+        qml.CNOT(wires=[0, 1])
+        return qml.var(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+    @qml.qnode(dev2, interface=interface)
+    def qnode2(x):
+        qml.Hadamard(wires=0)
+        qml.RX(x[0], wires=0)
+        qml.RY(x[1], wires=1)
+        qml.CNOT(wires=[0, 1])
+        return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(1))
+
+    return qnode1, qnode2
+
+
+class TestMap:
+    """Test for mapping ansatz over observables or devices,
+    to return a QNode cluster"""
+
+    def test_template_not_callable(self):
+        """test that an exception is correctly called if a
+        template is not callable"""
+        with pytest.raises(ValueError, match="template is not a callable"):
+            qml.map(5, 0, 0)
+
+    def test_mapping_over_observables(self):
+        """test that mapping over a list of observables produces
+        a QNodeCluster with the correct QNodes, with a single
+        device broadcast."""
+        dev = qml.device("default.qubit", wires=1)
+        obs_list = [qml.PauliX(0), qml.PauliY(0)]
+        template = lambda x, wires: qml.RX(x, wires=0)
+
+        qc = qml.map(template, obs_list, dev)
+
+        assert len(qc) == 2
+
+        # evaluate cluster so that queue is populated
+        qc(1)
+
+        assert len(qc[0].ops) == 2
+        assert qc[0].ops[0].name == "RX"
+        assert qc[0].ops[1].name == "PauliX"
+
+        assert len(qc[1].ops) == 2
+        assert qc[1].ops[0].name == "RX"
+        assert qc[1].ops[1].name == "PauliY"
+
+        # test that device is broadcast
+        assert qc[0].device is qc[1].device
+
```

Comment:
Extra test for ``tuples``:
```
    def test_mapping_over_observables_as_tuples(self):
        """Test that mapping over a tuple of observables produces
        a QNodeCluster with the correct QNodes, with a single
        device broadcast."""
        dev = qml.device("default.qubit", wires=1)
        obs_list = (qml.PauliX(0), qml.PauliY(0))
        template = lambda x, wires: qml.RX(x, wires=0)

        qc = qml.map(template, obs_list, dev)

        assert len(qc) == 2

        # evaluate cluster so that queue is populated
        qc(1)

        assert len(qc[0].ops) == 2
        assert qc[0].ops[0].name == "RX"
        assert qc[0].ops[1].name == "PauliX"

        assert len(qc[1].ops) == 2
        assert qc[1].ops[0].name == "RX"
        assert qc[1].ops[1].name == "PauliY"

        # test that device is broadcast
        assert qc[0].device is qc[1].device
```

Reply:
This is brilliant, thanks @antalszava!